<a href='http://www.algebra.hr'> <img src='../algebra_logo_color_h.png' alt="Algebra" width="500" /></a>
___
# Pandas manipulacija podacima

Do sada smo naučili kako učitati podatke i kako ih prilagoditi za daljnju obradu. Pokušajmonešto pročitati iz tih podataka. Koristit ćemo se podacima iz csv datoteka:<br>
- *Data\Vjezbe_Primjeri\airbnb_new_york_city_listings.csv* - sažetak podataka i mjerenja za ponudu na AirBnB u New Yorku
- *Data\Vjezbe_Primjeri\airbnb_new_york_city_reviews_full.csv* - podaci o ocjenama ponude na AirBnB u New Yorku
- *Data\Vjezbe_Primjeri\mall_customers.csv* - za potrebe dodatnog pojašnjavanja funkcionalnosti Pandas biblioteke - podaci su preuzeti s https://www.kaggle.com/kandij/mall-customers
- *Data\Vjezbe_Primjeri\ocjene.csv* - za potrebe dodatnog pojašnjavanja funkcionalnosti Pandas biblioteke - podaci su generirani nasumično pomoću web usluge https://www.json-generator.com/ preuzete s http://insideairbnb.com/get-the-data.html.

Za početak, pripremimo ih za analizu što ćemo iskoristiti kao malo ponavljanje prvog dijela gradiva o Pandas biblioteci.

In [2]:
import pandas as pd

abnb_ny_listings_df = pd.read_csv('../Data/Vjezbe_Primjeri/airbnb_new_york_city_listings.csv')

In [3]:
abnb_ny_listings_df

id                                               name   host_id  \
0          2595                              Skylit Midtown Castle      2845   
1          3831  Whole flr w/private bdrm, bath & kitchen(pls r...      4869   
2          5121                                    BlissArtsSpace!      7356   
3          5136           Spacious Brooklyn Duplex, Patio + Garden      7378   
4          5178                   Large Furnished Room Near B'way　      8967   
...         ...                                                ...       ...   
36719  51447515         Gorgeous 2 Bedroom - in Prime Midtown East  51589519   
36720  51449962  Charming UWS 1-bdrm, 1-office, 1/2 block to Ct...   2971741   
36721  51450816                                    The Hunter IIII  61391963   
36722  51451029                        Rockaway Beach Surf Getaway  10123226   
36723  51451368                       Nyc apt in the middle in nyc  17770287   

            host_name neighbourhood_group       neighbourhood   latitude  \
0            Jennifer           Manhattan             Midtown  40.753560   
1         LisaRoxanne            Brooklyn  Bedford-Stuyvesant  40.684940   
2               Garon            Brooklyn  Bedford-Stuyvesant  40.685350   
3             Rebecca            Brooklyn         Sunset Park  40.662650   
4            Shunichi           Manhattan             Midtown  40.764570   
...               ...                 ...                 ...        ...   
36719             Asi           Manhattan             Midtown  40.755787   
36720      Dina Marie           Manhattan     Upper West Side  40.787029   
36721  Stay With Vibe           Manhattan     Upper East Side  40.768950   
36722           Chase              Queens             Arverne  40.599257   
36723            Nina           Manhattan         Murray Hill  40.748655   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
0     -73.985590  Entire home/apt    150              30                 48   
1     -73.957650  Entire home/apt     75               1                408   
2     -73.955120     Private room     60              30                 50   
3     -73.994540  Entire home/apt    275               5                  1   
4     -73.983170     Private room     61               2                485   
...          ...              ...    ...             ...                ...   
36719 -73.965126  Entire home/apt    255              31                  0   
36720 -73.969239  Entire home/apt    306               5                  0   
36721 -73.960455  Entire home/apt     63              30                  0   
36722 -73.797953     Private room     75               1                  0   
36723 -73.981209  Entire home/apt    115              30                  0   

      last_review  reviews_per_month  calculated_host_listings_count  \
0      2019-11-04               0.34                               3   
1      2021-06-29               5.09                               1   
2      2016-06-05               0.55                               1   
3      2014-01-02               0.01                               1   
4      2021-07-18               3.63                               1   
...           ...                ...                             ...   
36719         NaN                NaN                               6   
36720         NaN                NaN                               1   
36721         NaN                NaN                              96   
36722         NaN                NaN                               1   
36723         NaN                NaN                              11   

       availability_365  
0                   341  
1                   212  
2                   365  
3                   184  
4                   255  
...                 ...  
36719               364  
36720                16  
36721               334  
36722                87  
36723               364  

[36724 rows 

Prvo ćemo se upoznati s tipovima podataka u kolonama. Oni će nam pomoći u slučaju kada neki tip podatka trebamo promijeniti i kada želimo popuniti prazne ćelije podacima. Primjer je kolona "Name", koja je tekstualnog tipa i svako ćemo prazno polje popuniti, tako što ćemo upisati vrijednost "Nedostaje ime", a za numeričke tipove ćemo dodati vrijednost 0.00 ili samo 0. Sve navedeno čini veliku razliku.

In [5]:
abnb_ny_listings_df.dtypes

id                                  int64
name                               object
host_id                             int64
host_name                          object
neighbourhood_group                object
neighbourhood                      object
latitude                          float64
longitude                         float64
room_type                          object
price                               int64
minimum_nights                      int64
number_of_reviews                   int64
last_review                        object
reviews_per_month                 float64
calculated_host_listings_count      int64
availability_365                    int64
dtype: object

Kolona "last_review" je object tip podataka, a u toj koloni su datumi zadnjih ocjena. Promijenimo tip podataka u toj koloni, u *DateTime* tipu podatka, pomoću metode *to_datetime()* koja kao argument prima naziv kolone.

*INFORMACIJA: mogli bismo koristiti i naš "stari" način pomoću astype() metode kojoj kao argument predamo naziv tipa podatka. Primjer prikazan unutar komentara.*

Kolona "price" je cjelobrojnog tipa podataka, a prikladniji je decimalni pa promijenimo tip i toj koloni.

In [9]:
# abnb_ny_listings_df['last_review'] = abnb_ny_listings_df['last_review'].astype('datetime')

abnb_ny_listings_df['last_review'] = pd.to_datetime(abnb_ny_listings_df['last_review'])
abnb_ny_listings_df['price'] = abnb_ny_listings_df['price'].astype('float')

abnb_ny_listings_df.dtypes

id                                         int64
name                                      object
host_id                                    int64
host_name                                 object
neighbourhood_group                       object
neighbourhood                             object
latitude                                 float64
longitude                                float64
room_type                                 object
price                                    float64
minimum_nights                             int64
number_of_reviews                          int64
last_review                       datetime64[ns]
reviews_per_month                        float64
calculated_host_listings_count             int64
availability_365                           int64
dtype: object

Sada kada imamo tip podatka *DateTime*, to možemo iskoristiti za neke dodatne akcije nad DF-om, kao što su na primjer dodavanje nove kolone naziva "last_review_year" u kojoj će biti pohranjeni podaci o godini zadnje ocjene, ali ne i cijeli datum.

In [15]:
abnb_ny_listings_df['last_review_year'] = abnb_ny_listings_df['last_review'].dt.year
# abnb_ny_listings_df['last_review_year'] = abnb_ny_listings_df['last_review_year'].astype('int')

abnb_ny_listings_df.dtypes

id                                         int64
name                                      object
host_id                                    int64
host_name                                 object
neighbourhood_group                       object
neighbourhood                             object
latitude                                 float64
longitude                                float64
room_type                                 object
price                                    float64
minimum_nights                             int64
number_of_reviews                          int64
last_review                       datetime64[ns]
reviews_per_month                        float64
calculated_host_listings_count             int64
availability_365                           int64
last_review_year                         float64
dtype: object

Sada imamo novu kolonu, ali su u njoj pohranjeni podaci u decimalnom zapisu. Godine trebaju biti zapisane kao cijeli brojevi. To ćemo riješiti nakon provjere toga  imamo li prazne ćelije.

Mi smo koristili *empty* svojstvo, ali ćemo pomoću metode dobiti malo više podataka o praznim ćelijama.

In [18]:
abnb_ny_listings_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36724 entries, 0 to 36723
Data columns (total 17 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   id                              36724 non-null  int64         
 1   name                            36711 non-null  object        
 2   host_id                         36724 non-null  int64         
 3   host_name                       36701 non-null  object        
 4   neighbourhood_group             36724 non-null  object        
 5   neighbourhood                   36724 non-null  object        
 6   latitude                        36724 non-null  float64       
 7   longitude                       36724 non-null  float64       
 8   room_type                       36724 non-null  object        
 9   price                           36724 non-null  float64       
 10  minimum_nights                  36724 non-null  int64         
 11  nu

In [17]:
abnb_ny_listings_df.isnull().sum()

id                                   0
name                                13
host_id                              0
host_name                           23
neighbourhood_group                  0
neighbourhood                        0
latitude                             0
longitude                            0
room_type                            0
price                                0
minimum_nights                       0
number_of_reviews                    0
last_review                       9415
reviews_per_month                 9415
calculated_host_listings_count       0
availability_365                     0
last_review_year                  9415
dtype: int64

In [19]:
abnb_ny_listings_df.dtypes

id                                         int64
name                                      object
host_id                                    int64
host_name                                 object
neighbourhood_group                       object
neighbourhood                             object
latitude                                 float64
longitude                                float64
room_type                                 object
price                                    float64
minimum_nights                             int64
number_of_reviews                          int64
last_review                       datetime64[ns]
reviews_per_month                        float64
calculated_host_listings_count             int64
availability_365                           int64
last_review_year                         float64
dtype: object

Imamo pet kolona koje imaju prazne ćelije. Kolone "name" i "host_name" su tipa object, odnosno tekstualnog tipa, dok su "reviews_per_month" i "last_review_year" decimalni brojevi, a kolona "last_review" je tipa *DateTime*.

Unesimo podatke u prazne ćelije tako da nam kasnije ne utječu na proračune. To će najbolje biti ako za kolonu:<br>
- "name" unesemo vrijednost "No Name Listing"
- "host_name" unesemo vrijednost "Bez imena" ili engleski uvriježeni oblik za bezimene "Jane Doe" za ženske osobe ili "John Doe" za muške osobe, nadalje mi ćemo koristi "J. Doe" da ostanemo neutralni
- "reviews_per_month" unesemo vrijednost 0.00 jer broj ocjena mjesečno može biti decimalni broj
- "last_review_year" unesemo vrijednost 0
- "last_review" unesemo vrijednost '01-01-1970' što bi odgovaralo početku UNIX vremena

Za ovu aktivnost imamo na raspolaganju tri metode:<br>
- *fillna()*
- *replace()*
- *interpolate()*

Mi ćemo koristiti metodu *fillna()*.

*INFORMACIJA: više detalja o navedenim funkcijama možete pronaći na stranici: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.fillna.html.*

In [23]:
abnb_ny_listings_df['name'] = abnb_ny_listings_df['name'].fillna('No Name Listing')

abnb_ny_listings_df['host_name'] = abnb_ny_listings_df['host_name'].fillna('J. Doe')
abnb_ny_listings_df['last_review'] = abnb_ny_listings_df['last_review'].fillna('1970-01-01')
abnb_ny_listings_df['last_review_year'] = abnb_ny_listings_df['last_review_year'].fillna(1970.0)
abnb_ny_listings_df['reviews_per_month'] = abnb_ny_listings_df['reviews_per_month'].fillna(0.0)

Provjerimo ponovno koliko je ćelija koje nemaju vrijednost.

In [24]:
abnb_ny_listings_df.isnull().sum()

id                                0
name                              0
host_id                           0
host_name                         0
neighbourhood_group               0
neighbourhood                     0
latitude                          0
longitude                         0
room_type                         0
price                             0
minimum_nights                    0
number_of_reviews                 0
last_review                       0
reviews_per_month                 0
calculated_host_listings_count    0
availability_365                  0
last_review_year                  0
dtype: int64

Odlično, nema niti jedna prazna ćelija. Provjerimo opet tipove podataka jer je moguće da smo ovom promjenom napravili i neke druge koje nismo željeli.

In [25]:
abnb_ny_listings_df.dtypes

id                                         int64
name                                      object
host_id                                    int64
host_name                                 object
neighbourhood_group                       object
neighbourhood                             object
latitude                                 float64
longitude                                float64
room_type                                 object
price                                    float64
minimum_nights                             int64
number_of_reviews                          int64
last_review                       datetime64[ns]
reviews_per_month                        float64
calculated_host_listings_count             int64
availability_365                           int64
last_review_year                         float64
dtype: object

Kolona "last_review" ponovno je tekstualni tip podatka, a "last_review_year" kolonu smo ionako željeli prebaciti u cjelobrojni tip umjesto decimalnog. Napravimo to za ove dvije kolone.

In [26]:
abnb_ny_listings_df['last_review_year'] = abnb_ny_listings_df['last_review_year'].astype('int')

abnb_ny_listings_df.dtypes

id                                         int64
name                                      object
host_id                                    int64
host_name                                 object
neighbourhood_group                       object
neighbourhood                             object
latitude                                 float64
longitude                                float64
room_type                                 object
price                                    float64
minimum_nights                             int64
number_of_reviews                          int64
last_review                       datetime64[ns]
reviews_per_month                        float64
calculated_host_listings_count             int64
availability_365                           int64
last_review_year                           int32
dtype: object

Sada je već bolje. Prikažimo naš DF da vidimo što smo napravili.

In [27]:
abnb_ny_listings_df

id                                               name   host_id  \
0          2595                              Skylit Midtown Castle      2845   
1          3831  Whole flr w/private bdrm, bath & kitchen(pls r...      4869   
2          5121                                    BlissArtsSpace!      7356   
3          5136           Spacious Brooklyn Duplex, Patio + Garden      7378   
4          5178                   Large Furnished Room Near B'way　      8967   
...         ...                                                ...       ...   
36719  51447515         Gorgeous 2 Bedroom - in Prime Midtown East  51589519   
36720  51449962  Charming UWS 1-bdrm, 1-office, 1/2 block to Ct...   2971741   
36721  51450816                                    The Hunter IIII  61391963   
36722  51451029                        Rockaway Beach Surf Getaway  10123226   
36723  51451368                       Nyc apt in the middle in nyc  17770287   

            host_name neighbourhood_group       neighbourhood   latitude  \
0            Jennifer           Manhattan             Midtown  40.753560   
1         LisaRoxanne            Brooklyn  Bedford-Stuyvesant  40.684940   
2               Garon            Brooklyn  Bedford-Stuyvesant  40.685350   
3             Rebecca            Brooklyn         Sunset Park  40.662650   
4            Shunichi           Manhattan             Midtown  40.764570   
...               ...                 ...                 ...        ...   
36719             Asi           Manhattan             Midtown  40.755787   
36720      Dina Marie           Manhattan     Upper West Side  40.787029   
36721  Stay With Vibe           Manhattan     Upper East Side  40.768950   
36722           Chase              Queens             Arverne  40.599257   
36723            Nina           Manhattan         Murray Hill  40.748655   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
0     -73.985590  Entire home/apt  150.0              30                 48   
1     -73.957650  Entire home/apt   75.0               1                408   
2     -73.955120     Private room   60.0              30                 50   
3     -73.994540  Entire home/apt  275.0               5                  1   
4     -73.983170     Private room   61.0               2                485   
...          ...              ...    ...             ...                ...   
36719 -73.965126  Entire home/apt  255.0              31                  0   
36720 -73.969239  Entire home/apt  306.0               5                  0   
36721 -73.960455  Entire home/apt   63.0              30                  0   
36722 -73.797953     Private room   75.0               1                  0   
36723 -73.981209  Entire home/apt  115.0              30                  0   

      last_review  reviews_per_month  calculated_host_listings_count  \
0      2019-11-04               0.34                               3   
1      2021-06-29               5.09                               1   
2      2016-06-05               0.55                               1   
3      2014-01-02               0.01                               1   
4      2021-07-18               3.63                               1   
...           ...                ...                             ...   
36719  1970-01-01               0.00                               6   
36720  1970-01-01               0.00                               1   
36721  1970-01-01               0.00                              96   
36722  1970-01-01               0.00                               1   
36723  1970-01-01               0.00                              11   

       availability_365  last_review_year  
0                   341              2019  
1                   212              2021  
2                   365              2016  
3                   184              2014  
4                   255              2021  
...                 ...               ...  
36719            

Odlično, imamo jednu kolonu više od inicijalnog skupa podataka, nemamo praznih ćelija.

Većinu tipova podataka smo očistili, međutim, često se dogodi da nam tekstualni podaci sadrže znakove koje računala ne prikazuju (primjer je znak za novi red '*\n*'). Ti znakovi su tu, no kako nam ne trebaju u podacima, bilo bi ih dobro obrisati. Pandas podržava string metodu *str.strip()* nad *DataFrame* objektom. Ova metoda briše sve posebne znakove, prazna mjesta (razmaknicu ili *space*) te slične znakove na početku i na kraju tekstualnog podatka.

Očistimo i tekstualne podatke pomoću *str.strip()* metode, u svakoj koloni koja ima tekstualni tip podatka.

*INFORMACIJA: možemo koristiti i str.lower() i str.upper() metode ako nam trebaju. U našem slučaju one nisu korisne. Više informacija možete pronaći na stranici: https://pandas.pydata.org/docs/reference/api/pandas.Series.str.strip.html.*

In [28]:
abnb_ny_listings_df.dtypes

id                                         int64
name                                      object
host_id                                    int64
host_name                                 object
neighbourhood_group                       object
neighbourhood                             object
latitude                                 float64
longitude                                float64
room_type                                 object
price                                    float64
minimum_nights                             int64
number_of_reviews                          int64
last_review                       datetime64[ns]
reviews_per_month                        float64
calculated_host_listings_count             int64
availability_365                           int64
last_review_year                           int32
dtype: object

In [29]:
abnb_ny_listings_df['name'] = abnb_ny_listings_df['name'].str.strip()
abnb_ny_listings_df['host_name'] = abnb_ny_listings_df['host_name'].str.strip()
abnb_ny_listings_df['neighbourhood_group'] = abnb_ny_listings_df['neighbourhood_group'].str.strip()
abnb_ny_listings_df['neighbourhood'] = abnb_ny_listings_df['neighbourhood'].str.strip()
abnb_ny_listings_df['room_type'] = abnb_ny_listings_df['room_type'].str.strip()

In [30]:
abnb_ny_listings_df

id                                               name   host_id  \
0          2595                              Skylit Midtown Castle      2845   
1          3831  Whole flr w/private bdrm, bath & kitchen(pls r...      4869   
2          5121                                    BlissArtsSpace!      7356   
3          5136           Spacious Brooklyn Duplex, Patio + Garden      7378   
4          5178                    Large Furnished Room Near B'way      8967   
...         ...                                                ...       ...   
36719  51447515         Gorgeous 2 Bedroom - in Prime Midtown East  51589519   
36720  51449962  Charming UWS 1-bdrm, 1-office, 1/2 block to Ct...   2971741   
36721  51450816                                    The Hunter IIII  61391963   
36722  51451029                        Rockaway Beach Surf Getaway  10123226   
36723  51451368                       Nyc apt in the middle in nyc  17770287   

            host_name neighbourhood_group       neighbourhood   latitude  \
0            Jennifer           Manhattan             Midtown  40.753560   
1         LisaRoxanne            Brooklyn  Bedford-Stuyvesant  40.684940   
2               Garon            Brooklyn  Bedford-Stuyvesant  40.685350   
3             Rebecca            Brooklyn         Sunset Park  40.662650   
4            Shunichi           Manhattan             Midtown  40.764570   
...               ...                 ...                 ...        ...   
36719             Asi           Manhattan             Midtown  40.755787   
36720      Dina Marie           Manhattan     Upper West Side  40.787029   
36721  Stay With Vibe           Manhattan     Upper East Side  40.768950   
36722           Chase              Queens             Arverne  40.599257   
36723            Nina           Manhattan         Murray Hill  40.748655   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
0     -73.985590  Entire home/apt  150.0              30                 48   
1     -73.957650  Entire home/apt   75.0               1                408   
2     -73.955120     Private room   60.0              30                 50   
3     -73.994540  Entire home/apt  275.0               5                  1   
4     -73.983170     Private room   61.0               2                485   
...          ...              ...    ...             ...                ...   
36719 -73.965126  Entire home/apt  255.0              31                  0   
36720 -73.969239  Entire home/apt  306.0               5                  0   
36721 -73.960455  Entire home/apt   63.0              30                  0   
36722 -73.797953     Private room   75.0               1                  0   
36723 -73.981209  Entire home/apt  115.0              30                  0   

      last_review  reviews_per_month  calculated_host_listings_count  \
0      2019-11-04               0.34                               3   
1      2021-06-29               5.09                               1   
2      2016-06-05               0.55                               1   
3      2014-01-02               0.01                               1   
4      2021-07-18               3.63                               1   
...           ...                ...                             ...   
36719  1970-01-01               0.00                               6   
36720  1970-01-01               0.00                               1   
36721  1970-01-01               0.00                              96   
36722  1970-01-01               0.00                               1   
36723  1970-01-01               0.00                              11   

       availability_365  last_review_year  
0                   341              2019  
1                   212              2021  
2                   365              2016  
3                   184              2014  
4                   255              2021  
...                 ...               ...  
36719            

Provjerimo jesmo li napravili neke neplanirane promjene.

In [31]:
abnb_ny_listings_df.dtypes

id                                         int64
name                                      object
host_id                                    int64
host_name                                 object
neighbourhood_group                       object
neighbourhood                             object
latitude                                 float64
longitude                                float64
room_type                                 object
price                                    float64
minimum_nights                             int64
number_of_reviews                          int64
last_review                       datetime64[ns]
reviews_per_month                        float64
calculated_host_listings_count             int64
availability_365                           int64
last_review_year                           int32
dtype: object

In [32]:
abnb_ny_listings_df.isnull().sum()

id                                0
name                              0
host_id                           0
host_name                         0
neighbourhood_group               0
neighbourhood                     0
latitude                          0
longitude                         0
room_type                         0
price                             0
minimum_nights                    0
number_of_reviews                 0
last_review                       0
reviews_per_month                 0
calculated_host_listings_count    0
availability_365                  0
last_review_year                  0
dtype: int64

Napravimo još jednu preinaku u podacima kako bismo krenuli s uporabom podataka. Za početak prikažimo još jednom sve kolone:

In [33]:
abnb_ny_listings_df.columns

Index(['id', 'name', 'host_id', 'host_name', 'neighbourhood_group',
       'neighbourhood', 'latitude', 'longitude', 'room_type', 'price',
       'minimum_nights', 'number_of_reviews', 'last_review',
       'reviews_per_month', 'calculated_host_listings_count',
       'availability_365', 'last_review_year'],
      dtype='object')

Vidimo da imamo kolone naziva "price" i "minimum_nights". Jedna kolona prikazuje cijenu najma po noćenju, a druga prikazuje koliko se minimalno noćenja mora uzeti. Ako spojimo ove dvije kolone, možemo dobiti novu koja će nam prikazati podatak o minimalnom prihodu. Dakle, ako pomnožimo cijenu s minimalnim brojem noćenja, onda je to minimalni prihod koji iznajmljivač može dobiti za svoju nekretninu.

Ako nova kolona prikazuje podatak o minimalnom prihodu, onda ćemo je nazvati "minimal_revenue". Napravimo to.

In [34]:
abnb_ny_listings_df['minimal_revenue'] = abnb_ny_listings_df['price'] * abnb_ny_listings_df['minimum_nights']

abnb_ny_listings_df

id                                               name   host_id  \
0          2595                              Skylit Midtown Castle      2845   
1          3831  Whole flr w/private bdrm, bath & kitchen(pls r...      4869   
2          5121                                    BlissArtsSpace!      7356   
3          5136           Spacious Brooklyn Duplex, Patio + Garden      7378   
4          5178                    Large Furnished Room Near B'way      8967   
...         ...                                                ...       ...   
36719  51447515         Gorgeous 2 Bedroom - in Prime Midtown East  51589519   
36720  51449962  Charming UWS 1-bdrm, 1-office, 1/2 block to Ct...   2971741   
36721  51450816                                    The Hunter IIII  61391963   
36722  51451029                        Rockaway Beach Surf Getaway  10123226   
36723  51451368                       Nyc apt in the middle in nyc  17770287   

            host_name neighbourhood_group       neighbourhood   latitude  \
0            Jennifer           Manhattan             Midtown  40.753560   
1         LisaRoxanne            Brooklyn  Bedford-Stuyvesant  40.684940   
2               Garon            Brooklyn  Bedford-Stuyvesant  40.685350   
3             Rebecca            Brooklyn         Sunset Park  40.662650   
4            Shunichi           Manhattan             Midtown  40.764570   
...               ...                 ...                 ...        ...   
36719             Asi           Manhattan             Midtown  40.755787   
36720      Dina Marie           Manhattan     Upper West Side  40.787029   
36721  Stay With Vibe           Manhattan     Upper East Side  40.768950   
36722           Chase              Queens             Arverne  40.599257   
36723            Nina           Manhattan         Murray Hill  40.748655   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
0     -73.985590  Entire home/apt  150.0              30                 48   
1     -73.957650  Entire home/apt   75.0               1                408   
2     -73.955120     Private room   60.0              30                 50   
3     -73.994540  Entire home/apt  275.0               5                  1   
4     -73.983170     Private room   61.0               2                485   
...          ...              ...    ...             ...                ...   
36719 -73.965126  Entire home/apt  255.0              31                  0   
36720 -73.969239  Entire home/apt  306.0               5                  0   
36721 -73.960455  Entire home/apt   63.0              30                  0   
36722 -73.797953     Private room   75.0               1                  0   
36723 -73.981209  Entire home/apt  115.0              30                  0   

      last_review  reviews_per_month  calculated_host_listings_count  \
0      2019-11-04               0.34                               3   
1      2021-06-29               5.09                               1   
2      2016-06-05               0.55                               1   
3      2014-01-02               0.01                               1   
4      2021-07-18               3.63                               1   
...           ...                ...                             ...   
36719  1970-01-01               0.00                               6   
36720  1970-01-01               0.00                               1   
36721  1970-01-01               0.00                              96   
36722  1970-01-01               0.00                               1   
36723  1970-01-01               0.00                              11   

       availability_365  last_review_year  minimal_revenue  
0                   341              2019           4500.0  
1                   212              2021             75.0  
2                   365              2016           1800.0  
3                   184              2014           1375.0  
4                   

In [35]:
abnb_ny_listings_df.dtypes

id                                         int64
name                                      object
host_id                                    int64
host_name                                 object
neighbourhood_group                       object
neighbourhood                             object
latitude                                 float64
longitude                                float64
room_type                                 object
price                                    float64
minimum_nights                             int64
number_of_reviews                          int64
last_review                       datetime64[ns]
reviews_per_month                        float64
calculated_host_listings_count             int64
availability_365                           int64
last_review_year                           int32
minimal_revenue                          float64
dtype: object

In [36]:
abnb_ny_listings_df

id                                               name   host_id  \
0          2595                              Skylit Midtown Castle      2845   
1          3831  Whole flr w/private bdrm, bath & kitchen(pls r...      4869   
2          5121                                    BlissArtsSpace!      7356   
3          5136           Spacious Brooklyn Duplex, Patio + Garden      7378   
4          5178                    Large Furnished Room Near B'way      8967   
...         ...                                                ...       ...   
36719  51447515         Gorgeous 2 Bedroom - in Prime Midtown East  51589519   
36720  51449962  Charming UWS 1-bdrm, 1-office, 1/2 block to Ct...   2971741   
36721  51450816                                    The Hunter IIII  61391963   
36722  51451029                        Rockaway Beach Surf Getaway  10123226   
36723  51451368                       Nyc apt in the middle in nyc  17770287   

            host_name neighbourhood_group       neighbourhood   latitude  \
0            Jennifer           Manhattan             Midtown  40.753560   
1         LisaRoxanne            Brooklyn  Bedford-Stuyvesant  40.684940   
2               Garon            Brooklyn  Bedford-Stuyvesant  40.685350   
3             Rebecca            Brooklyn         Sunset Park  40.662650   
4            Shunichi           Manhattan             Midtown  40.764570   
...               ...                 ...                 ...        ...   
36719             Asi           Manhattan             Midtown  40.755787   
36720      Dina Marie           Manhattan     Upper West Side  40.787029   
36721  Stay With Vibe           Manhattan     Upper East Side  40.768950   
36722           Chase              Queens             Arverne  40.599257   
36723            Nina           Manhattan         Murray Hill  40.748655   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
0     -73.985590  Entire home/apt  150.0              30                 48   
1     -73.957650  Entire home/apt   75.0               1                408   
2     -73.955120     Private room   60.0              30                 50   
3     -73.994540  Entire home/apt  275.0               5                  1   
4     -73.983170     Private room   61.0               2                485   
...          ...              ...    ...             ...                ...   
36719 -73.965126  Entire home/apt  255.0              31                  0   
36720 -73.969239  Entire home/apt  306.0               5                  0   
36721 -73.960455  Entire home/apt   63.0              30                  0   
36722 -73.797953     Private room   75.0               1                  0   
36723 -73.981209  Entire home/apt  115.0              30                  0   

      last_review  reviews_per_month  calculated_host_listings_count  \
0      2019-11-04               0.34                               3   
1      2021-06-29               5.09                               1   
2      2016-06-05               0.55                               1   
3      2014-01-02               0.01                               1   
4      2021-07-18               3.63                               1   
...           ...                ...                             ...   
36719  1970-01-01               0.00                               6   
36720  1970-01-01               0.00                               1   
36721  1970-01-01               0.00                              96   
36722  1970-01-01               0.00                               1   
36723  1970-01-01               0.00                              11   

       availability_365  last_review_year  minimal_revenue  
0                   341              2019           4500.0  
1                   212              2021             75.0  
2                   365              2016           1800.0  
3                   184              2014           1375.0  
4                   

Sada trebamo sve ovo ponoviti za skup podataka o ocjenama nekretnina, za skup podataka ocjena korisnika trgovačkog lanca i ocjene studenata.

Krenimo s ocjenama nekretnina. Ovaj skup podataka ima kolonu s datumima pa ćemo ovaj put konverziju datuma napraviti prilikom učitavanja podataka izravno u *read_csv()* metodi, koja prima dva argumenta vezana uz konverziju datuma. To su:<br>
- *parse_dates*
- *date_parser.*

Više detalja na: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html.

Mi ćemo koristiti *parse_dates* argument tako što ćemo tom argumentu proslijediti listu sa samo jednim elementom, odnosno nazivom kolone.

**VJEŽBA**<br>
**Probajte ovu opciju tako da argumentu parse_dates proslijedite Boolean vrijednost *True parse_dates=True.***

In [44]:
abnb_ny_reviews_df = pd.read_csv('../Data/Vjezbe_Primjeri/airbnb_new_york_city_reviews_full.csv', parse_dates=['date'])
abnb_ny_reviews_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 839321 entries, 0 to 839320
Data columns (total 6 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   listing_id     839321 non-null  int64         
 1   id             839321 non-null  int64         
 2   date           839321 non-null  datetime64[ns]
 3   reviewer_id    839321 non-null  int64         
 4   reviewer_name  839313 non-null  object        
 5   comments       838492 non-null  object        
dtypes: datetime64[ns](1), int64(3), object(2)
memory usage: 38.4+ MB


Imamo šest kolona i 839321 redaka.

Provjerimo koji su tipovi podataka u ovih 6 kolona.

In [45]:
abnb_ny_reviews_df.dtypes

listing_id                int64
id                        int64
date             datetime64[ns]
reviewer_id               int64
reviewer_name            object
comments                 object
dtype: object

Provjerom tipa podataka po kolonama, vidimo da kolona za datum ima ispravan tip podatka tako da nema potrebe za njenom konverzijom. Nakon provjere imamo li praznih ćelija, trebat ćemo samo *počistiti* tekstualne podatke u kolonama "reviewer_name" i "comments".

Provjerimo imamo li praznih ćelija i koliko ih je.

In [46]:
abnb_ny_reviews_df.isnull().sum()

listing_id         0
id                 0
date               0
reviewer_id        0
reviewer_name      8
comments         829
dtype: int64

Ćelije bez podataka imamo u kolonama "comments" i "reviewer_name". Zato ćemo prvo popuniti te ćelije te nakon toga "počistiti" tekstualne podatke.

Sada više nemamo praznih ćelija. Vrijeme je da *očistimo* tekstualne podatke.

**VJEŽBA**<br>
**Napravite skupove podataka s ocjenama korisnika trgovačkog centra te ocjenama učenika.**

Vjerujem da ste svi došli do ovoga rješenja rješavajući zadatak iz uvodnog dijela o Pandas biblioteci (osim dijela o popunjavanju praznih ćelija). Ako niste, nije problem, tijekom ovog ponavljanja trebali ste doći na dio koji ćemo koristiti u nastavku.

## Manipulacija podacima - traženje informacija u podacima

Sada kada smo naše podatke pripremili možemo se "igrati" podacima.

Na početku ćemo napraviti nekoliko koraka koji bi mogli biti uključeni u pripremu podataka, ali i manipulaciju podacima. U svakom slučaju, to su koraci koji mogu pomoći u procesu "traženja informacija u podacima".

Prvi takav korak je uklanjanje "viška" kolona. Često skupovi podataka imaju jako puno kolona pa se onda teško usredotočiti na neku skupinu kolona koje nam mogu dati tražene informacije. Osim toga, ponekad postoje kolone koje je sustav generirao tijekom prikupljanja podataka, ali isključivo kao potporu sustavu tijekom procesa prikpljanja podataka. Takve kolone ne predstavljaju nikakvu korist tijekom obrade podataka pa ih je onda korisno maknuti iz skupa podataka.

Metoda koju ćemo koristiti je *drop()* (detalji o metodi iz Pandas dokumentacije: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop.html?highlight=drop#pandas.DataFrame.drop).

Ova metoda može se koristiti za brisanje, ne samo kolona, nego i redaka. Prvi argument je lista naziva kolone ili kolona koje želimo izbrisati. Naziv argumenta je "labels". Uz ovaj argument,potrebno je navesti i osi (axis) unutar kojih želimo brisati podatke.

*Primjer: dataFrame.drop('naziv_kolone', axis=1).*

Iz gornjeg primjera, vidimo da će se izbrisati element čiji smo naziv proslijedili, a da radi se o koloni koju definira os 1 (axis=1).

Alternativu gornjem primjeru možemo dobiti korištenjem argumenta *columns=\['naziv_kolone', ...\] ili index=\['naziv_indeksa', ...\].*

**Obratite pažnju na to da indeksi NE moraju nužno biti numerički, nego mogu biti i drugačijih zapisa.**

**VJEŽBA: uklonite kolone: *'id', 'name', 'host_id', 'host_name'***.

**PITANJE **<br>
**U tablici opet imamo kolone 'latitude', 'longitude'. To su kolone koje smo maknuli.  Kako ćemo to riješiti?**

ODGOVOR: koristit ćemo *Argument inplace - bool, default False -> inplace=True.*

Maknimo "trajno" kolone.

Provjerimo jesmo li sada uklonili željene kolone.

**Promjena naziva kolona**

Ako dohvatimo listu kolona unutar skupa podataka "ocjene_df" (*ocjene_df.columns*) dobit ćemo ovaj rezultat: *Index(\['fname', 'lname', 'gender', 'age', 'exercise', 'hours', 'grade', 'address'\], dtype='object')*.

- 'fname' - predstavlja ime učenika/ce
- 'lname' - predstavlja prezime učenika/ce
- 'gender' - predstavlja spol učenika/ce
- 'age' - predstavlja dob učenika/ce
- 'exercise' - predstavlja tjedni broj sati vježbanja učenika/ce
- 'hours' - predstavlja tjedni broj sati učenja učenika/ce
- 'grade'- predstavlja ocjenu učenika/ce
- 'address'- predstavlja adresu učenika/ce

Za ovaj skup podataka počeli smo koristiti hrvatske nazive pa prevedimo onda i nazive kolona u skladu s gore navedenim opisima.

Promjenu naziva kolone možemo napraviti na više načina.

**Način 1.**

Pomoću atributa columns koji ima svaki *DataFrame* objekt. Jednostavno u listu kolona navedemo nove vrijednosti i to je to:
*ocjene_df.columns = \['Ime', 'Prezime', 'Spol', 'Dob', 'Sati_Tjednog_Vjezbanja', 'Sati_Tjednog_Ucenja', 'Ocjena', 'Puna_Adresa'\].*

**Način 2.**

Pomoću atributa columns i Python liste.

*Korak 1.*<br>
Kreirajmo listu od naziva kolona.<br>
*kolone = list(ocjene_df.columns.values)*

*Korak 2.*<br>
Unesemo željene izmjene.<br>
*kolone\[0\] = 'Ime'*<br>
*kolone\[1\] = 'Prezime'*<br>
*kolone\[2\] = 'Spol'*<br>
*kolone\[3\] = 'Dob'*<br>
*kolone\[4\] = 'Sati_Tjednog_Vjezbanja'*<br>
*kolone\[5\] = 'Sati_Tjednog_Ucenja'*<br>
*kolone\[6\] = 'Ocjena'*<br>
*kolone\[7\] = 'Puna_Adresa'*

*Korak 3.*<br>
Pridružimo vrijednost liste kolekciji kolona skupa podataka.<br>
*ocjene_df.columns = kolone*

**Način 3.**

Pomoću metode *rename()*.
ocjene_df.rename(<br>
&nbsp;&nbsp;&nbsp;&nbsp;columns={<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;'fname': 'Ime',<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;'lname': 'Prezime',<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;'gender': 'Spol',<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;'age': 'Dob',<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;'exercise': 'Sati_Tjednog_Vjezbanja',<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;'hours': 'Sati_Tjednog_Ucenja',<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;'grade': 'Ocjena',<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;'address': 'Puna_Adresa'<br>
&nbsp;&nbsp;&nbsp;&nbsp;}<br>
)<br>

**Zadatak**<br>
**Napravite izmjene pomoću Načina 3. na skupu podataka *ocjene_df.***

**Provjerite nazive kolona za preostala dva skupa podataka te ako mislite da je potrebno, predložite nove nazive kolona.**

Ponekad trebamo dodatno prilagoditi tekst naziva kolona tako što će nam trebati sva mala ili sva velika slova u nazivu kolona. Jedan od primjera je potreba za kreiranjem mape za svaku kolonu pa je za Linux OS prikladno koristiti sva mala slova.

Znamo da su kolone unutar DataFrame objekta grupirane unutar liste pa iskoristimo Python za prolazak kroz svaki element liste i primjenimo odgovarajuću metodu nad tekstom naziva kolone.

**Način 1.**

Pomoću Python *map()* funkcije<br>
*ocjene_df.columns = map(str.lower, ocjene_df.columns)*

**Način 2.**
Pomoću Python *for* petlje.<br>
*ocjene_df.columns = \[x.lower() for x in ocjene_df.columns\]*

**Zadatak**<br>
**Napravite izmjene pomoću Načina 2. na skupu podataka *ocjene_df.***

**Na preostala dva skupa podataka prebacite nazive kolona u velika slova.**

**Filtriranje podataka**

Ponekad želimo izdvojiti ili dohvatiti određene podatke iz grupe. Ovisno o cilju, najčešće ćemo definirati neke uvjete. Kao i do sada, ovo možemo napraviti na više načina.

Započnimo korištenjem *str.contains()* metodom. Dohvatimo sve podatke o nekretninama iz Manhattana. Prvo ćemo provjeriti ima li bilo koji podatak koji zadovoljava ovaj uvjet:

Rezultat je *True*, dakle postoje takvi podaci u skupu podataka. Provjerimo jesu li svi podaci s ovim uvjetom.

Za ovaj skup podataka je očekivano, ali je ponekad korisno provjeriti jer ako svi podaci zadovoljavaju određeni uvjet, onda nema potrebe za filtriranjem. Probajmo dohvatiti u kojim redovima su podaci koji ispunjavaju traženi uvjet.

Dobili smo samo podatak ispunjava li redak uvjet ili ne. Međutim, za dohvat svih podataka u retku, trebamo se koristiti svojstvom *loc\[\]* koje kao argument prima listu Boolean vrijednosti. Napravimo to, a kao vrijednost argumenta  predat ćemo gornju liniju koda te dodati uvjet da je jednako *True*, kako nam ne bi vratio sve.

**Zadatak**<br>
**Iz skupa podataka o ocjenama, prikažite samo retke učenika s ocjenama većim od 70.
Dodajte filter i po spolu.**<br>
**VAŽNO: koristit ćemo znakove za logičke operacije iz C programskog jezika:**
- **& - and**
- **| - or.**

**Statistička obrada podataka**

Prošli smo filtriranje podataka i prije nego se upoznamo s grupiranjem, upoznajmo se sa statističkom obradom podataka. Statistika će nam pomoći da se još bolje upoznamo s našim podacima.

Za početak provjerimo najčešće korištene statističke vrijednosti:
- minimalna vrijednost (*min*)
- maksimalna vrijednost (*max*)
- raspon (*max - min*)
- srednja vrijednosti (*mean*)
- medijan (*median*) - vrijednost koja se nalazi u sredini nakon što podatke poredamo po veličini
- standardna devijacija (*std*)
- ukupan broj vrijednosti (*count*) na kolonama koje su nam najzanimljivije za ove podatke.

Za skup podataka o nekretninama to su kolone:
- *price*
- *minimum_nights*
- *number_of_reviews*
- *last_review*
- *reviews_per_month*
- *availability_365*
- *last_review_year*
- *minimal_revenue*

*INFORMACIJA: kada budemo koristili vizualizaciju podataka, onda ćemo ove vrijednosti prikazati grafički pa će biti puno lakše uočiti neke granične vrijednosti.*

**VJEŽBA**<br>
**Dovršite analizu za preostale kolone.**

Srećom, Pandas ima ugrađenu funkciju *describe()* koja objedinjuje većinu gore navedene statističke vrijednosti.

Možemo dodati i argument *include='all'* koji uključuje sve dostupne izračune.

*INFORMACIJA: dodajte traženi argument.*

Gore navedeni statistički izračuni vrijede za sve kolone ili za sve retke unutar neke kolone (serije). Primjenimo stečeno znanje o statističkim izračunima na ono što smo već naučili o filtriranju podataka.

**Sortiranje i grupiranje podataka**

Često podatke trebamo sortirati po veličini, od najmanje prema najvećoj ili obrnuto. To radimo pomoću metode *sort_values()* koja ima dva najvažnija parametra:
- *by* - listu kolona koje želimo sortirati i
- *ascending* - koji određuje je li sortiranje od manje prema većoj vrijednosti ili obrnuto.

Rekli smo da možemo sortirati više kolona. Jednostavno navedimo više kolona kao vrijednost *by* parametra.

**ZADATAK**<br>
**Sortirajte jednu ili više kolona iz grupa podataka o nekretninama i njihovim ocjenama (*listings_df*, *reviews_df*).**

**Grupiranje podataka**

Podatke često grupiramo po nekoj kategoriji. Na primjer, svi apartmani ili svi učenici neke dobi. Za tu grupu podataka onda napravimo određeni statistički izračun (srednja vrijednost, medijan i sl.).

Iz primjera niže, prvo smo izdvojili tri kolone, od kojih smo onda napravili grupiranje po tipu nekretnine i za svaku od tih grupa smo izračunali srednju vrijednost.

*INFORMACIJA: ukoliko ste "izgubili" kolonu 'minimal_revenue', a može se dogoditi jer radimo u jednom velikom Python programu, ponovite njezino kreiranje. Ovo se može dogoditi zato što radimo dosta primjera, često pokrećemo različite ćelije i to ne nužno onim redoslijedom kako smo ih napisali pa zbog toga može doći do grešaka. U stvarnom životu, kada radite s jednom skupinom podataka, odmah je razdvojite na manje cjeline i onda na svakoj manjoj cjelini radite zasebno.*

Podatke možemo grupirati, ne samo u jednoj koloni, nego u više njih.

**ZADATAK**<br>
**Grupirajte jednu ili više kolona iz grupa podataka o ocjenama učenika.**

**Korelacija između podataka u kolonama**

Često ćemo željeti otkriti u kakvoj su vezi podaci iz jedne kolone s podacima iz druge, odnosno postoji li nekakva korelacija između njih. Na primjer, broj sati učenja i ocjena mogu biti u korelaciji, ali ime i ocjena sigurno ne bi trebali imati međusobni utjecaj, odnosno korelaciju.

Pandas ima ugrađenu metodu nad *DataFrame* objektom naziva *corr()*. Ova metoda vraća vrijednosti do 1 do -1. Ako je korelacija između dviju kolona visoka, onda će vrijednost biti 1 ili jako blizu jedinice, a ako kolone nisu u korelaciji, onda će vrijednost biti negativna, odnosno bit će blizu minus jedan.

Pogledajmo kako *corr()* metoda radi na skupu podataka s ocjenama učenika.

U tabeli se vidi da najveću vrijednost (0.801955) imaju kolone ocjena i *sati_tjednog_ucenja*. Zatim kolone *ocjena* i *sati_tjednog_vjezbanja*.

**ZADATAK**<br>
**Pronađite korelacije za skupine podataka o nekretninama i njihovim ocjenama (*listings_df*, *reviews_df*).**

**PIVOT tabele**

Pivot tabele ste vjerojatno upoznali u Microsoft Excelu. Njihova primjena je sumirati, sortirati, reorganizirati, grupirati, prebrojati sve ili prosjek podataka u tabeli.

Možda ovo pomogne boljem razumijevanju pivot tabele. Pivot tabela može:
- grupirati pojedinačne zapise, redove u kategorije
- prebrojati elemente u svakoj kategoriji
- sumirati vrijednosti elemenata
- izračunati srednju vrijednost, minimum ili maksimum.

Pandas ima ugrađenu metodu *pivot_table()*. Upoznajmo se s njezinim mogućnostima. Više detalja na: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.pivot_table.html?highlight=pivot_table#pandas.pivot_table.

Srećom, nema puno kolona, ali zamislimo da ima pa ćemo odabrati prikaz samo prosječne ocjene po spolu i dobi.

Bilo bi dobro kada bismo umjesto prosječne vrijednosti dobili najmanju vrijednost ocjene po spolu.

**ZADATAK**<br>
**Očigledno argument *aggfunc* predstavlja vrijednost funkcije po kojoj se agregira, odnosno grupira, odnosno predstavlja statističku vrijednost koju računamo. Probajte mijenjati *aggfunc* argument nazivima statističkih metoda u Pandasu te uz kolonu ocjene, dodajte i sate tjednog učenja (otkrili smo da su u koleraciji).**

Bilo bi korisno kada bismo na kraju *pivot* tabele, kao zadnji redak, imali i sumarni prikaz za sve vrijednosti određene kolone. Recimo, prosječna vrijednost za ocjene i sate tjednog učenja po spolu, ali i ukupna prosječna vrijednosti za ocjene i sate tjednog učenja BEZ obzira na spol.

### ZADACI ZA VJEŽBANJE

Analizirajte csv datoteke:

- *cars.csv*
- *e-cars.csv*
- *earthquakes.csv*

tako da pomoću Pandas metoda (*corr(), pivot_table() ...*) izvučete neke zanimljive informacije.